config/src/test/java/org/axonframework/config/SimpleStatefulCommandHandlingComponentTest.java

```diff
@@ -0,0 +1,323 @@
+        verifyStudentEnrolledInCourse("my-studentId-3", "my-courseId-1");
+        verifyStudentEnrolledInCourse("my-studentId-2", "my-courseId-1");
+
+        // Third and last possible student
+        enrollStudentToCourse(component, "my-studentId-4", "my-courseId-1");
+        verifyStudentEnrolledInCourse("my-studentId-4", "my-courseId-1");
+        verifyStudentEnrolledInCourse("my-studentId-3", "my-courseId-1");
+        verifyStudentEnrolledInCourse("my-studentId-2", "my-courseId-1");
+
+        // Fourth can still enroll for other course
+        enrollStudentToCourse(component, "my-studentId-4", "my-courseId-2");
+        verifyStudentEnrolledInCourse("my-studentId-4", "my-courseId-2");
+
+        // But five can not enroll for the first course
+        var exception = assertThrows(CompletionException.class,
+                     () -> enrollStudentToCourse(component, "my-studentId-5", "my-courseId-1"
+                     ));
+        assertInstanceOf(IllegalArgumentException.class, exception.getCause());
+        assertTrue(exception.getCause().getMessage().contains("Course already has 3 students"));
+    }
+
+    private void appendEvent(ProcessingContext context, Object event) {
+        eventStore.transaction(context, DEFAULT_CONTEXT)
+                  .appendEvent(new GenericEventMessage<>(
+                          new MessageType(event.getClass()),
+                          event));
+    }
+
+    private void verifyStudentEnrolledInCourse(String id, String courseId) {
+        AsyncUnitOfWork uow = new AsyncUnitOfWork();
+        uow.executeWithResult(context -> studentRepository
+                .load(id, context)
+                .thenAccept(student -> assertTrue(student.entity().getCoursesEnrolled().contains(courseId)))
+                .thenCompose(v -> courseRepository.load(courseId, context))
+                .thenAccept(course -> assertTrue(course.entity().getStudentsEnrolled().contains(id))))
+           .join();
+    }
+
+    private void updateStudentName(StatefulCommandHandlingComponent component, String id, String name) {
+        sendCommand(component, new ChangeStudentNameCommand(id, name));
+    }
+
+
+    private void enrollStudentToCourse(
+            StatefulCommandHandlingComponent component,
+            String studentId,
+            String courseId
+    ) {
+        sendCommand(component, new EnrollStudentToCourseCommand(studentId, courseId));
+    }
+
+    private <T> void sendCommand(
+            StatefulCommandHandlingComponent component,
+            T payload
+    ) {
+        AsyncUnitOfWork uow = new AsyncUnitOfWork();
+        uow.executeWithResult(context -> {
+            GenericCommandMessage<T> command = new GenericCommandMessage<>(
+                    new MessageType(payload.getClass()),
+                    payload);
+            return component.handle(command, context).first().asCompletableFuture();
+        }).join();
+    }
+
+
+    record ChangeStudentNameCommand(
```

Comment:
Not something to deal with in this PR, but just sharing it as food for thought.
What about having some test domain in Axon Framework that we use throughout our tests?

AF4 currently has numerous instantiations of dummy commands, event, queries, aggregates, sagas, and projections. While if we have one sample domain inside **this** project, we could (try to) make that into the "super domain" supporting all functionality.

Anyhow, I think that would be nice. Picking the student-and-course combo for that seems, to be honest, a good one to me.

Reply:
Totally agree! Good idea.

Reply:
Yes, I've been thinking about this. Putting **all** functionality into **one** domain might be complex though. But I think we can cover 80% easily.

Reply:
As you see, I have added this test domain :) Resolving comment